pkg/server/image_pull.go

```diff
@@ -331,13 +331,29 @@ func (c *criContainerdService) pullImage(ctx context.Context, rawRef string, aut
 	// Use config digest as imageID to conform to oci image spec, and also add image id as
 	// image reference.
 	imageID := configDesc.Digest.String()
-	if err := c.imageStoreService.Put(ctx, imageID, desc); err != nil {
-		return "", "", "", fmt.Errorf("failed to put image id %q into containerd image store: %v",
-			imageID, err)
+	if err := c.createImageReference(ctx, imageID, desc); err != nil {
+		return "", "", "", fmt.Errorf("failed to update image id %q: %v", imageID, err)
 	}
 	return imageID, repoTag, repoDigest, nil
 }
 
+// createImageReference creates image reference inside containerd image store.
+func (c *criContainerdService) createImageReference(ctx context.Context, name string, desc imagespec.Descriptor) error {
+	img := containerdimages.Image{
+		Name:   name,
+		Target: desc,
+	}
+	_, err := c.imageStoreService.Create(ctx, img)
```

Comment:
For performance reasons I recommend we go with update then create. See: https://github.com/kubernetes-incubator/cri-containerd/pull/101/files#diff-87348f253685e1fb4a3238134a550238R307

My assumption is more updates (positive cache hits) than creates (negative cache hits).